### PR TITLE
Object retention put/get APIs

### PR DESCRIPTION
**Pull request by prakashsvmx**
Object retention APIs.

Note: this depends on #919 

<summary> Test Script </summary>

<details>

var Minio = require("../../dist/main/minio")

var minioClient = new Minio.Client({
  endPoint:"127.0.0.1",
  port:9000,
  useSSL:false,
  accessKey:"minio",
  secretKey :"minio123"
})
const file = '/home/prakash/Downloads/HouseDesign/E00000000833.jpg'

function makeBucketWithObjectLock (buckName,makeOpts,cb){
  const bucketTaggingPromise = minioClient.makeBucket(buckName,makeOpts)
  bucketTaggingPromise.then(() => {
    console.log("makeBucketWithObjectLock success...")
    cb && cb()
  })
    .catch((e)=>{
      console.log("makeBucketWithObjectLock Error",e)
    })
}

function setObjectLockConfig (buckName,lockOpts,cb){
  const bucketTaggingPromise = minioClient.setObjectLockConfig(buckName,lockOpts)
  bucketTaggingPromise.then(() => {
    console.log("setObjectLockConfig success...")
    cb && cb()
  })
    .catch((e)=>{
      console.log("setObjectLockConfig Error",e)
    })

}

function getObjectLockConfig (buckName,cb){
  const bucketTaggingPromise = minioClient.getObjectLockConfig(buckName)
  bucketTaggingPromise.then((e, data) => {
    console.log("getObjectLockConfig success...", e, JSON.stringify(data))
    cb && cb()
  })
    .catch((e)=>{
      console.log("getObjectLockConfig Error",e)
    })
}

function putObject (buckName,objName) {
  var metaData = {
    'Content-Type': 'application/octet-stream',
    'X-Amz-Meta-Testing': 1234,
    'example': 5678
  }

  const fPromise = minioClient.fPutObject(buckName, objName, file,metaData)

  fPromise.then(() => {
    console.log("Put object Success...")
  })
    .catch(()=>{
      console.log("Put object  Error")
    })

}

function putObjectRetention (buckName,objectName, retentionOpts,cb){

  const bucketTaggingPromise = minioClient.putObjectRetention(buckName,objectName,retentionOpts)
  bucketTaggingPromise.then(() => {
    console.log("Promise working...")
    cb && cb()
  })
    .catch((e)=>{
      console.log("Promise Error",e)
    })

}

function getObjectRetention(buckName,objectName,getOpts){

  const bucketTaggingPromise = minioClient.getObjectRetention(buckName,objectName,getOpts, )
  bucketTaggingPromise.then((data) => {
    console.log("Get retention working...", data)
  })
    .catch((e)=>{
      console.log("Promise Error",e)
    })
}

function removeObject (bucketName, objectName, removeOpts){
  const delObj= minioClient.removeObject(bucketName, objectName, removeOpts)
  delObj.then(()=>{
    console.log("Removed object")
  })
}

function removeBucket (bucketName){
  const delObj= minioClient.removeBucket(bucketName)
  delObj.then(()=>{
    console.log("Removed bucket")
  })
}
const bucketName = 'sph-my-bucket'
const objName ="E00000000833"

const expirationDate = new Date('2021-08-02')
expirationDate.setUTCHours(0,0,0,0)//Should be start of the day.(midnight)
const versionId ="OjYzf_hoYoAfeEYO3Ymc8_bToDH6F0Cb"
//makeBucketWithObjectLock(bucketName,{ObjectLocking:true})
//putObject(bucketName,objName)
/*putObjectRetention(bucketName,objName,{
    governanceBypass:true,
    mode:"GOVERNANCE",
    retainUntilDate:expirationDate.toISOString(),
    versionId:versionId
})*/
//getObjectRetention(bucketName,objName,{versionId:versionId})
//removeObject(bucketName,objName,{versionId:versionId, governanceBypass:true})
//removeBucket(bucketName)

</details>

### CODE REVIEW

**Review by ebozduman — CHANGES_REQUESTED** [review]

**Comment by prakashsvmx** [comment]
@ebozduman thank you so much for the review. I will address them .

File: examples/put-object-retention.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+
+const bucketName = 'my-bucket'
+const objectName ="my-object"
+
+const expirationDate = new Date('2021-08-02')
+expirationDate.setUTCHours(0,0,0,0)//Should be start of the day.(midnight)
+const versionId ="e67b4b08-144d-4fc4-ba15-43c3f7f9ba74"
+
+const objRetPromise = s3Client.putObjectRetention(bucketName,objectName,{governanceBypass:true,Mode:"GOVERNANCE",retainUntilDate:expirationDate.toISOString(),versionID:versionId} )
```
[the remark this answers is not in the document]
Comment (prakashsvmx) [reply examples/put-object-retention.js]: @ebozduman  The examples has typos (case sensitivity) of object keys. I have updated and tested against MinIO and S3. 
however. still there is a response code mismatch as commented on the code. 

Line 2426: (minio.js)

`    //FIXME minio Server returns 204 but AWS returns 200. So in aws, though the operation is success,  error is thrown.
`

File: docs/API.md [hunk 2]
```diff
@@ -1047,6 +1063,69 @@ minioClient.removeIncompleteUpload('mybucket', 'photo.jpg', function(err) {
...
+    function (err, res){
+    if (err) {
+        return console.log(err)
+    }
+    console.log("Success", res)
+})
+```
+
+<a name="getObjectRetention"></a>
+### getObjectRetention(bucketName, objectName, getOpts [,callback])
```
[the remark this answers is not in the document]
Comment (prakashsvmx) [reply docs/API.md]: The `getOpts`  will have only `versionId` so i have updated.

File: docs/API.md [hunk 3]
```diff
@@ -1047,6 +1063,69 @@ minioClient.removeIncompleteUpload('mybucket', 'photo.jpg', function(err) {
...
+
+Apply retention on an object.
+
+__Parameters__
+
+| Param  |  Type | Description  |
+|---|---|---|
+| `bucketName`  |_string_   | Name of the bucket.  |
+| `objectName`  | _string_  | Name of the object.  |
+| `retentionOpts` | _object_ | Options for retention like : `{ governanceBypass:true/false ,mode:COMPLIANCE/GOVERNANCE, retainUntilDate: _date_ , versionId:<object_version_uuid> }`  Default is `{}`|
```
[the remark this answers is not in the document]
Comment (prakashsvmx) [reply docs/API.md]: I have updated it for consistency.

File: docs/API.md [hunk 4]
```diff
@@ -933,18 +935,32 @@ minioClient.removeObject('mybucket', 'photo.jpg', function(err) {
...
+minioClient.removeObject('mybucket', 'photo.jpg', { versionId : "my-version-uuid" }, function(err) {
   if (err) {
     return console.log('Unable to remove object', err)
   }
   console.log('Removed the object')
 })
 ```
 
+__Example 3__
+Remove an object version with retention option (governanceBypass)
```
[the remark this answers is not in the document]
Comment (prakashsvmx) [reply docs/API.md]: Thank you for the suggestion. I have applied the same.

**Review by ebozduman — APPROVED** [review]
LGTM

File: src/test/functional/functional-tests.js [hunk 5]
```diff
@@ -1754,4 +1754,101 @@ describe('functional tests', function() {
...
+        }else{
+          done()
+        }
+
+      })
+
+      step(`putObjectRetention(bucketName, objectName, putOpts)_bucketName:${objRetentionBucket}, objectName:${retentionObjName}`, done => {
+        //Put two versions of the same object.
+        if(isFeatureSupported) {
+          const expirationDate = new Date('2021-08-02')
```
Comment (kannappanr) [inline src/test/functional/functional-tests.js:1803]: Can you get the current time and add a say or so, instead of hard coding the date here?

Comment (prakashsvmx) [reply src/test/functional/functional-tests.js]: Thank you :+1:  . I have updated the date to be dynamic.

**Review by kannappanr — APPROVED** [review]
LGTM

File: examples/put-object-retention.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  endPoint: 's3.amazonaws.com',
+  accessKey: 'YOUR-ACCESSKEYID',
+  secretKey: 'YOUR-SECRETACCESSKEY'
+})
+
+
+const bucketName = 'my-bucket'
+const objectName ="my-object"
+
+const expirationDate = new Date('2021-08-02')
```
Comment (ebozduman) [inline examples/put-object-retention.js:32]: You have one more hard-coded date here...

**Review by ebozduman — CHANGES_REQUESTED** [review]

File: examples/put-object-retention.js (hunk 6 above)
Comment (prakashsvmx) [reply examples/put-object-retention.js:32]: Thanks a lot. I have updated here and in API example as well.

**Review by kannappanr — APPROVED** [review]
LGTM

**Review by ebozduman — APPROVED** [review]
LGTM

**Pull request merged**